Guard rocket_defer_inline_exclusions filter return when used by 3rd parties

engahmeds3ed opened the pull request:
## Description

I did two things here on this PR related to the return of `rocket_defer_inline_exclusions` filter:-
1. I casted its return to be always array.
2. For backward compatibility, I checked if the return of filter is string so I converted it to array by slicing it using separator `|`

Fixes #3576 

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] This change requires a documentation update

## Is the solution different from the one proposed during the grooming?

Yes It's like groomed solution.

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Add the following code into active theme `functions.php` and then clear the cache and check `error_log` or `debug.log`
```
add_filter( 'rocket_defer_inline_exclusions', function( $exclusions ){
    return 'any_parameter';
} );
```
OR
```
add_filter( 'rocket_defer_inline_exclusions', function( $exclusions ){
    return [ 'any_parameter', 'second_parameter' ];
} );
```
OR
```
add_filter( 'rocket_defer_inline_exclusions', function( $exclusions ){
    return new StdClass();
} );
```

- [x] By using ShortPixel AI plugin as in the issue description.

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

engahmeds3ed committed:
guard filter return value to be array and for backward compatibility to convert string to array

engahmeds3ed committed:
add testcases for string, array and object filter returns

remyperona commented:
@engahmeds3ed Is the PR ready for review? It's still in draft for now.

engahmeds3ed commented:
> @engahmeds3ed Is the PR ready for review? It's still in draft for now.

Yes it's ready for review.
I didn't notice that it's draft, changed now.

engahmeds3ed commented:
@wp-media/qa ready for QA

piotrbak commented:
@engahmeds3ed Before we were getting those errors with Smart Slider:
https://snippi.com/s/eznltbk

After this update, the notice `Array to string conversion in` in their plugin is still present. Is it expected?

Their code:
https://snippi.com/s/0keypyy

This is the 56th line:
`$regexp .= '|';`

engahmeds3ed commented:
@piotrbak 
Yes we are guarding our filter return not to let any plugin return any type of data from this filter so here we expect array and loop on it, so if any code returned string from this filter, our code will give the following warning
```
PHP Warning:  Invalid argument supplied for foreach() in /wp-content/plugins/wp-rocket/inc/Engine/Optimization/DeferJS/DeferJS.php on line 141
```

but the first warning on your snippet is from smartslider code so they need to adjust it.

piotrbak commented:
@engahmeds3ed When Smart Slider or ShortPixel are enabled, this is how the `$inline_exclusions` looks like:
`[2021-02-19 19:20:41] $inline_exclusions: Array|N2R|spai_settings`
https://github.com/wp-media/wp-rocket/blob/b8f237caa52d2dfe4de305ea2c0d25c17e41696d/inc/Engine/Optimization/DeferJS/DeferJS.php#L149

Do we expect a multidimensional array there?

engahmeds3ed commented:
@piotrbak 

No, they are trying to append string data to array so it appears like what you sent.
we expect array and guarded it from being changed.

piotrbak commented:
Our exclusions are not working when their plugins are enabled:
![image](https://user-images.githubusercontent.com/26767037/108554553-2398d900-72ec-11eb-9131-37264dd9ab8d.png)

Is that something we can take care of on our end @engahmeds3ed ?

Before this update, the `$inline_exclusions` was empty, causing nothing to be enclosed in the DomContentLoaded.

Mai-Saad commented:
@engahmeds3ed Please check this case as it results in a fatal error (not sure if it`s a case to happen in reality):
Add the following code into active theme functions.php and then clear the cache and check error_log or debug.log
```
add_filter( 'rocket_defer_inline_exclusions', function( $exclusions ){
    return [ 1, 2];
} );
```

engahmeds3ed committed:
handle array of integers filter values

engahmeds3ed committed:
use our exclusions list when inline exclusions list is empty

engahmeds3ed commented:
@piotrbak Yes we can handle this case, I added a check to fill the `$inline_exclusions` with the default list pattern when the filter return is empty or not valid.

@Mai-Saad I adjusted the code now to handle array of any type.

engahmeds3ed committed:
change default inline exclusions to keep our list away from any change

engahmeds3ed committed:
fix formatting

engahmeds3ed commented:
Now I changed the way we deal with this filter so this filter return will be added to our main list (no changing it anymore) and make the filter default value to be null so it can accept any time and then do the type check to convert this additional exclusions from string to be array (if it's string) or cast it to array (if it's not an array).

@wp-media/php Please validate before sending again to @wp-media/qa

crystinutzaa commented on inc/Engine/Optimization/DeferJS/DeferJS.php:
```diff
@@ -289,4 +274,36 @@ public function exclude_jquery_upgrade() {
+
+		$inline_exclusions_list = array_merge( $inline_exclusions_list, (array) $additional_inline_exclusions_list );
+
+		foreach ( (array) $inline_exclusions_list as $inline_exclusions_item ) {
```
I don't think type casted in there is needed.
`array_merge` function always return an array

crystinutzaa commented on inc/Engine/Optimization/DeferJS/DeferJS.php:
```diff
@@ -289,4 +274,36 @@ public function exclude_jquery_upgrade() {
+		 *
+		 * @param array $inline_exclusions_list Array of inline JS that should not be deferred.
+		 */
+		$additional_inline_exclusions_list = apply_filters( 'rocket_defer_inline_exclusions', null );
```
Why do you use `null` on apply_filters? 
`$additional_inline_exclusions_list` should be an array, right?
So why not using the default empty array?

engahmeds3ed replied:
as discussed with you on slack, we need to be compatible with both string and array types for backward compatibility.

crystinutzaa commented on inc/Engine/Optimization/DeferJS/DeferJS.php:
```diff
@@ -289,4 +274,36 @@ public function exclude_jquery_upgrade() {
+
+		$inline_exclusions = '';
+
+		if ( is_string( $additional_inline_exclusions_list ) ) {
```
Wouldn't be better to guard agaist `! is_array()`? 
And always `$additional_inline_exclusions_list` at this point to be converted to array?
Will reduce the numerous type castings from below.

engahmeds3ed replied:
this will be good with all types except for objects so when casting object to string and this object didn't implement the magic method `__toString` it will give an error but if u tried to cast this object to array it will work.

crystinutzaa commented on tests/Fixtures/inc/Engine/Optimization/DeferJS/Subscriber/deferInlineJs.php:
```diff
@@ -143,4 +143,126 @@ function newContent() {
+
+	'testShouldExcludeUsingStringFilter' => [
+		'config' => [
+			'rocket_defer_inline_exclusions_filter' => 'first_string|third_string',
```
What about adding a Bool (true / false), Int, Float / double, Null exclusion?
Should we add these types to fixtures?

engahmeds3ed committed:
adjust type checking beased on code review

engahmeds3ed committed:
add more test cases for filter return types

engahmeds3ed commented:
@wp-media/qa ready for another round of QA

piotrbak commented:
@engahmeds3ed Tested scenarios and results:

```
add_filter( 'rocket_defer_inline_exclusions', function( $exclusions ){
    $exclusions[] = "test1";
    $exclusions[] = "test2";
    return $exclusions;
} );
```
1. Works when plugins are disabled
2. When plugins are enabled, it adds `Array` to the `$inline_exclusions` and we're not excluding the pattern 🐛 

```
add_filter( 'rocket_defer_inline_exclusions', function( $exclusions ){
    return "test3";
} );
```
1. Works when plugins are disabled
2. Works when plugins are enabled

```
add_filter( 'rocket_defer_inline_exclusions', function( $exclusions ){
    $exclusions .= "test4|test5";
    return $exclusions;
} );
```
1. Works when plugins are disabled
2. Works when plugins are enabled

piotrbak commented:
@engahmeds3ed @Tabrisrp As we talked, it's ready to go. However, the most reliable way to exclude more than one attribute from DeferJS, that will also work with the current compatibility of Smart Slider3 + ShortPixel Adaptive Images is this one:
```
add_filter( 'rocket_defer_inline_exclusions', function( $exclusions ){
    $exclusions .= "test4|test5";
    return $exclusions;
} );
```

piotrbak reviewed (approved):
Worked as expected:
[testrail-report-138.pdf](https://github.com/wp-media/wp-rocket/files/6045593/testrail-report-138.pdf)
